File: cpp/trex_selector_methods/trex_da/demo_trex_da_01_mc_sim_ar1/demo_trex_da_01_mc_sim_ar1.cpp
// ==============================================================================
// demo_trex_da_01_mc_sim_ar1.cpp
// ==============================================================================
/**
 * @file demo_trex_da_01_mc_sim_ar1.cpp
 *
 * @brief DA-TRex Monte Carlo simulations for the AR(1) Toeplitz DGP.
 *
 * @details
 *  Part 1: MC simulation sweeping SNR (n=300, p=1000, s=10, rho=0.7, 200 MC,
 *          3 solvers: TLARS, TAFS, TOMP).
 *
 *  Part 2: MC simulation sweeping rho (n=300, p=1000, s=10, SNR=2.0, 200 MC,
 *          3 solvers: TLARS, TAFS, TOMP).
 *
 *  Part 3: 2D Gap × Rho sweep. Explores the kappa-boundary where
 *          gap < kappa collapses TPP. Support is generated via
 *          CappedSpread(gap) and Random policies. 3 solvers: TLARS, TAFS, TOMP.
 *
 *          gap_grid : {100, 50, 20, 15, 10, 5, 1}
 *          rho_grid : {0.0, 0.1, ..., 0.9}
 *
 * Results are saved to simulation_results/ as TXT and CSV files.
 */
// ==============================================================================

#include "trex_da_sim_common.hpp"

// ==============================================================================

using namespace da_sim;

// ==============================================================================
//  Part 1: MC simulation — SNR sweep, 2 support types (CappedSpread + Random)
// ==============================================================================

void demo_ar1_mc_snr_sweep()
{
    // Simulation config
    SimConfig cfg;
    cfg.n = 300;
    cfg.p = 1000;
    cfg.s = 10;
    cfg.amplitude = 3.0;
    cfg.tFDR = 0.2;
    constexpr int max_gap = 20;
    constexpr double rho = 0.7;

    TRexDAControlParameter da_ctrl;
    da_ctrl.method = DAMethod::AR1;

    const std::vector<double> snr_grid = {0.1, 0.2, 0.5, 1.0, 2.0, 5.0};
    const std::string hdr_base =
        "AR(1) rho=" + fmt_num(rho) +
        ", n=" + std::to_string(cfg.n) +
        ", p=" + std::to_string(cfg.p);


    // --- Run A: CappedSpread(max_gap=20) ---
    run_snr_sweep(
        "da_ar1_snr_capped",
        snr_grid,
        cfg.num_MC,
        cfg.tFDR,
        cfg.base_seed,
        default_solvers(),
        [&](double snr, unsigned seed) {
            return dgp_ar1(cfg.n,
                           cfg.p,
                           make_support(
                               SupportPolicy::CappedSpread,
                               cfg.s, cfg.p, max_gap, seed),
                           cfg.amplitude, rho, snr, seed);
        },
        da_ctrl,
        make_trex_control(cfg.K),
        hdr_base + ", support=CappedSpread(max_gap=" +
                      std::to_string(max_gap) + ")",
        /*include_base_trex=*/true);


    // --- Run B: Random support (redrawn per trial) ---
    run_snr_sweep(
        "da_ar1_snr_random",
        snr_grid,
        cfg.num_MC,
        cfg.tFDR,
        cfg.base_seed,
        default_solvers(),
        [&](double snr, unsigned seed) {
            return dgp_ar1(cfg.n,
                           cfg.p,
                           make_support(
                               SupportPolicy::Random,
                               cfg.s, cfg.p, 0, seed),
                           cfg.amplitude, rho, snr, seed);
        },
        da_ctrl,
        make_trex_control(cfg.K),
        hdr_base + ", support=Random (redrawn per trial)",
        /*include_base_trex=*/true);
}


// ==============================================================================
//  Part 2: MC simulation — Rho sweep, 2 support types (CappedSpread + Random)
// ==============================================================================

void demo_ar1_mc_rho_sweep()
{
    // Simulation config
    SimConfig cfg;
    cfg.n = 300;
    cfg.p = 1000;
    cfg.s = 10;
    cfg.amplitude = 3.0;
    cfg.tFDR = 0.2;
    constexpr int    max_gap   = 20;
    constexpr double snr_fixed = 2.0;

    TRexDAControlParameter da_ctrl;
    da_ctrl.method = DAMethod::AR1;

    const std::vector<double> rho_grid = {
        0.0, 0.1, 0.2, 0.3, 0.4,
        0.5, 0.6, 0.7, 0.8, 0.9
    };
    const std::string hdr_base =
        "AR(1) SNR=" + fmt_num(snr_fixed) +
        ", n=" + std::to_string(cfg.n) +
        ", p=" + std::to_string(cfg.p);


    // --- Run A: CappedSpread(max_gap=20) ---
    run_param_sweep(
        "da_ar1_rho_capped",
        "Rho",
        rho_grid,
        cfg.num_MC,
        cfg.tFDR,
        cfg.base_seed,
        default_solvers(),
        [&](double rho_val, unsigned seed) {
            return dgp_ar1(cfg.n,
                           cfg.p,
                           make_support(
                               SupportPolicy::CappedSpread,
                               cfg.s, cfg.p, max_gap, seed),
                           cfg.amplitude, rho_val, snr_fixed, seed);
        },
        da_ctrl,
        make_trex_control(cfg.K),
        hdr_base + ", support=CappedSpread(max_gap=" +
                     std::to_string(max_gap) + ")",
        /*include_base_trex=*/true);

    // --- Run B: Random support (redrawn per trial) ---
    run_param_sweep(
        "da_ar1_rho_random",
        "Rho",
        rho_grid,
        cfg.num_MC,
        cfg.tFDR,
        cfg.base_seed,
        default_solvers(),
        [&](double rho_val, unsigned seed) {
            return dgp_ar1(cfg.n, cfg.p,
                           make_support(SupportPolicy::Random,
                                        cfg.s, cfg.p, 0, seed),
                           cfg.amplitude, rho_val, snr_fixed, seed);
        },
        da_ctrl,
        make_trex_control(cfg.K),
        hdr_base + ", support=Random (redrawn per trial)",
        /*include_base_trex=*/true);
}


// ==============================================================================
//  Part 3: 2D Gap × Rho sweep
//
//  Runs num_MC trials per (solver, gap, rho) cell for two support types:
//    A) CappedSpread(gap) — deterministic, evenly-spaced support.
//    B) Random           — support redrawn each trial (not gap-dependent).
//
//  The kappa boundary (kappa = ceiling(log(0.02) / log(rho))) is the DA+AR1
//  correction window half-width.  When gap < kappa, active predictors fall
//  inside each other's correction windows and TPP collapses.
//
//  gap_grid : {100, 50, 20, 15, 10, 5, 1}
//  rho_grid : {0.0, 0.1, ..., 0.9}
// ==============================================================================

void demo_ar1_mc_gap_rho_sweep()
{
    SimConfig cfg;
    cfg.n      = 300;
    cfg.p      = 1000;
    cfg.s      = 10;
    cfg.num_MC = 200;
    cfg.amplitude = 3.0;
    cfg.tFDR = 0.2;
    constexpr double snr_fixed = 2.0;

    const std::vector<int> gap_grid = {100, 50, 20, 15, 10, 5, 1};
    const std::vector<double> rho_grid = {
        0.0, 0.1, 0.2, 0.3, 0.4,
        0.5, 0.6, 0.7, 0.8, 0.9
    };

    const auto n_gap = gap_grid.size();
    const auto n_rho = rho_grid.size();
    const auto n_rho_idx = static_cast<Eigen::Index>(n_rho);
    const auto n_gap_idx = static_cast<Eigen::Index>(n_gap);

    // Kappa boundary: ceil(log(0.02) / log(rho)).
    // For rho=0 (independent predictors) kappa=0 → gap always sufficient.
    std::vector<int> kappa_boundary(n_rho);
    for (std::size_t ir = 0; ir < n_rho; ++ir) {
        if (rho_grid[ir] <= 0.0)
            kappa_boundary[ir] = 0;
        else
            kappa_boundary[ir] = static_cast<int>(
                std::ceil(std::log(0.02) / std::log(rho_grid[ir])));
    }

    // Per-solver result maps
    std::map<std::string, Eigen::MatrixXd> tpp_cs_map,    fdp_cs_map,
                                           sd_tpp_cs_map,  sd_fdp_cs_map;
    std::map<std::string, Eigen::VectorXd> tpp_rand_map,   fdp_rand_map,
                                           sd_tpp_rand_map, sd_fdp_rand_map;

    const auto solvers = default_solvers();
    for (const auto& sv : solvers) {
        tpp_cs_map[sv.name]     = Eigen::MatrixXd::Zero(n_rho_idx, n_gap_idx);
        fdp_cs_map[sv.name]     = Eigen::MatrixXd::Zero(n_rho_idx, n_gap_idx);
        sd_tpp_cs_map[sv.name]  = Eigen::MatrixXd::Zero(n_rho_idx, n_gap_idx);
        sd_fdp_cs_map[sv.name]  = Eigen::MatrixXd::Zero(n_rho_idx, n_gap_idx);
        tpp_rand_map[sv.name]    = Eigen::VectorXd::Zero(n_rho_idx);
        fdp_rand_map[sv.name]    = Eigen::VectorXd::Zero(n_rho_idx);
        sd_tpp_rand_map[sv.name] = Eigen::VectorXd::Zero(n_rho_idx);
        sd_fdp_rand_map[sv.name] = Eigen::VectorXd::Zero(n_rho_idx);
    }

    TRexDAControlParameter da_ctrl;
    da_ctrl.method = DAMethod::AR1;

    const unsigned base_offset = static_cast<unsigned>(cfg.base_seed);

    // --- Header ---
    cdianostics::print_section_header("AR(1) Part 3: Gap × Rho sweep");
    std::cout << "  n=" << cfg.n << "  p=" << cfg.p << "  s=" << cfg.s
              << "  SNR=" << snr_fixed << "  MC=" << cfg.num_MC
              << "  tFDR=" << cfg.tFDR
              << "  amplitude=" << cfg.amplitude << "\n";
    std::cout << "  gap_grid: {";
    for (std::size_t i = 0; i < n_gap; ++i)
        std::cout << (i ? ", " : "") << gap_grid[i];
    std::cout << "}\n  rho_grid: {";
    for (std::size_t i = 0; i < n_rho; ++i)
        std::cout << (i ? ", " : "")
                  << std::fixed << std::setprecision(1) << rho_grid[i];
    std::cout << "}\n\n";

    // ---- Loop over solvers ----
    for (const auto& sv : solvers) {
        auto trex_ctrl = make_trex_control(cfg.K);
        trex_ctrl.solver_type           = sv.type;
        trex_ctrl.solver_params.rho_afs = sv.rho_afs;
        trex_ctrl.tloop_stagnation_stop = sv.use_stagnation;

        std::cout << "\n  Solver: " << sv.name << "\n\n";

        // ---- Section A: CappedSpread(gap) × rho ----
        std::cout << "  [A] CappedSpread(gap) sweep ...\n\n";
        int cell_idx = 0;
        const int total_cells = static_cast<int>(n_rho * n_gap);

        for (std::size_t ir = 0; ir < n_rho; ++ir) {
            const double rho_val = rho_grid[ir];
            for (std::size_t ig = 0; ig < n_gap; ++ig) {
                ++cell_idx;
                const int gap_val = gap_grid[ig];
                auto support = capped_spread_support(cfg.s, cfg.p, gap_val);

                std::ostringstream lbl;
                lbl << "[" << cell_idx << "/" << total_cells << "] "
                    << sv.name
                    << "  rho=" << std::fixed << std::setprecision(1) << rho_val
                    << "  gap=" << std::setw(3) << gap_val
                    << "  kappa=" << kappa_boundary[ir];

                auto res = run_mc_trials(
                    cfg.num_MC, lbl.str(),
                    [&](unsigned seed) {
                        return dgp_ar1(cfg.n, cfg.p, support,
                                       cfg.amplitude, rho_val, snr_fixed, seed);
                    },
                    cfg.tFDR, trex_ctrl, da_ctrl, base_offset);

                const auto ri = static_cast<Eigen::Index>(ir);
                const auto gi = static_cast<Eigen::Index>(ig);
                tpp_cs_map[sv.name](ri, gi)    = res.avg_tpr;
                fdp_cs_map[sv.name](ri, gi)    = res.avg_fdr;
                sd_tpp_cs_map[sv.name](ri, gi) = res.sd_tpr;
                sd_fdp_cs_map[sv.name](ri, gi) = res.sd_fdr;
            }
        }

        // ---- Section B: Random support sweep (rho only) ----
        std::cout << "\n  [B] Random support sweep ...\n\n";
        for (std::size_t ir = 0; ir < n_rho; ++ir) {
            const double rho_val = rho_grid[ir];

            std::ostringstream lbl;
            lbl << "[" << (ir + 1) << "/" << n_rho << "] "
                << sv.name
                << "  rho=" << std::fixed << std::setprecision(1) << rho_val
                << "  kappa=" << kappa_boundary[ir] << "  Random";

            auto res = run_mc_trials(
                cfg.num_MC, lbl.str(),
                [&](unsigned seed) {
                    return dgp_ar1(cfg.n, cfg.p,
                                   make_support(SupportPolicy::Random,
                                                cfg.s, cfg.p, 0, seed),
                                   cfg.amplitude, rho_val, snr_fixed, seed);
                },
                cfg.tFDR, trex_ctrl, da_ctrl, base_offset);

            const auto ri = static_cast<Eigen::Index>(ir);
            tpp_rand_map[sv.name](ri)    = res.avg_tpr;
            fdp_rand_map[sv.name](ri)    = res.avg_fdr;
            sd_tpp_rand_map[sv.name](ri) = res.sd_tpr;
            sd_fdp_rand_map[sv.name](ri) = res.sd_fdr;
        }
    }

    // ---- Save results ----
    const std::string header_extra =
        "n=" + std::to_string(cfg.n)  +
        ", p=" + std::to_string(cfg.p) +
        ", s=" + std::to_string(cfg.s) +
        ", SNR=" + fmt_num(snr_fixed) +
        ", tFDR=" + fmt_num(cfg.tFDR) +
        ", amplitude=" + fmt_num(cfg.amplitude);

    save_and_print_2d_gap_rho_results(
        "da_ar1_gap_rho",
        gap_grid,
        rho_grid,
        kappa_boundary,
        cfg.num_MC,
        solvers,
        tpp_cs_map,    fdp_cs_map,    sd_tpp_cs_map,  sd_fdp_cs_map,
        tpp_rand_map,  fdp_rand_map,  sd_tpp_rand_map, sd_fdp_rand_map,
        header_extra);
}


// ==============================================================================
//  main
// ==============================================================================

int main() {

    std::cout.setf(std::ios::unitbuf);
    omp_set_num_threads(6);
    std::cout << "Running with " << omp_get_max_threads() << " threads\n\n";

    if (false) demo_ar1_mc_snr_sweep();
    if (false) demo_ar1_mc_rho_sweep();
    if (true) demo_ar1_mc_gap_rho_sweep();

    std::cout << "\nAR(1) MC simulation complete.\n";
    return 0;
}

File: Python/trex_selector_methods/trex_da/demo_trex_da_01_mc_sim_ar1/demo_trex_da_01_mc_sim_ar1.py
# ==============================================================================
# demo_trex_da_01_mc_sim_ar1.py
# ==============================================================================
#
# DA-TRex+AR1 Monte Carlo simulations for the AR(1) Toeplitz DGP.
#
#   Part 1: MC SNR sweep (n=300, p=1000, s=10, rho=0.7, 200 MC; CappedSpread
#           and Random support; 3 DA solvers + base T-Rex comparison rows).
#   Part 2: MC rho sweep (SNR=2.0; same layout as Part 1).
#   Part 3: 2D Gap x Rho sweep — probes the kappa boundary where gap < kappa
#           collapses TPP (CappedSpread(gap) matrices + Random column).
#
# Mirrors cpp/trex_selector_methods/trex_da/demo_trex_da_01_mc_sim_ar1 (same
# DGP, control parameters, sweep axes, and result files). Results are saved to
# simulation_results/ as TXT and CSV files. Every part runs num_MC=200 trials
# per grid point — reduce NUM_MC to preview.
# ==============================================================================

import os
import sys

_THIS_DIR = os.path.dirname(os.path.abspath(__file__))
_PARENT_DIR = os.path.dirname(_THIS_DIR)
for _p in (_THIS_DIR, _PARENT_DIR):
    if _p not in sys.path:
        sys.path.insert(0, _p)

import numpy as np

from da_sim_common import (default_solvers, fmt_num, make_support_capped_spread,
                           print_section_header, run_mc_trials, run_param_sweep,
                           run_snr_sweep, save_and_print_2d_gap_rho_results)

RUN_PART_1 = True
RUN_PART_2 = True
RUN_PART_3 = True

OUT_DIR = os.path.join(_THIS_DIR, "simulation_results")

# Simulation config (mirrors cpp SimConfig + per-part constants)
N, P, S = 300, 1000, 10
AMPLITUDE = 3.0
TFDR = 0.2
K = 20
NUM_MC = 200
BASE_SEED = 2026
MAX_GAP = 20

DA_SPEC = {"method": "AR1"}


# ==============================================================================
# Part 1: MC simulation — SNR sweep, 2 support types (CappedSpread + Random)
# ==============================================================================

def part_1_snr_sweep():
    rho = 0.7
    snr_grid = [0.1, 0.2, 0.5, 1.0, 2.0, 5.0]
    hdr_base = f"AR(1) rho={fmt_num(rho)}, n={N}, p={P}"
    support_capped = make_support_capped_spread(S, P, MAX_GAP)

    # --- Run A: CappedSpread(max_gap=20) ---
    run_snr_sweep(
        OUT_DIR, "da_ar1_snr_capped", snr_grid, NUM_MC, TFDR, BASE_SEED,
        default_solvers("+AR1"),
        lambda snr: dict(n=N, p=P, dgp="dgp_ar1_snr",
                         kwargs=dict(amplitude=AMPLITUDE, rho=rho, snr=snr),
                         support=support_capped, s=S),
        DA_SPEC, K,
        hdr_base + f", support=CappedSpread(max_gap={MAX_GAP})",
        include_base_trex=True)

    # --- Run B: Random support (redrawn per trial) ---
    run_snr_sweep(
        OUT_DIR, "da_ar1_snr_random", snr_grid, NUM_MC, TFDR, BASE_SEED,
        default_solvers("+AR1"),
        lambda snr: dict(n=N, p=P, dgp="dgp_ar1_snr",
                         kwargs=dict(amplitude=AMPLITUDE, rho=rho, snr=snr),
                         support=None, s=S),
        DA_SPEC, K,
        hdr_base + ", support=Random (redrawn per trial)",
        include_base_trex=True)


# ==============================================================================
# Part 2: MC simulation — Rho sweep, 2 support types (CappedSpread + Random)
# ==============================================================================

def part_2_rho_sweep():
    snr_fixed = 2.0
    rho_grid = [round(0.1 * i, 1) for i in range(10)]   # 0.0 .. 0.9
    hdr_base = f"AR(1) SNR={fmt_num(snr_fixed)}, n={N}, p={P}"
    support_capped = make_support_capped_spread(S, P, MAX_GAP)

    # --- Run A: CappedSpread(max_gap=20) ---
    run_param_sweep(
        OUT_DIR, "da_ar1_rho_capped", "Rho", rho_grid, NUM_MC, TFDR, BASE_SEED,
        default_solvers("+AR1"),
        lambda rho: dict(n=N, p=P, dgp="dgp_ar1_snr",
                         kwargs=dict(amplitude=AMPLITUDE, rho=rho, snr=snr_fixed),
                         support=support_capped, s=S),
        DA_SPEC, K,
        hdr_base + f", support=CappedSpread(max_gap={MAX_GAP})",
        include_base_trex=True)

    # --- Run B: Random support (redrawn per trial) ---
    run_param_sweep(
        OUT_DIR, "da_ar1_rho_random", "Rho", rho_grid, NUM_MC, TFDR, BASE_SEED,
        default_solvers("+AR1"),
        lambda rho: dict(n=N, p=P, dgp="dgp_ar1_snr",
                         kwargs=dict(amplitude=AMPLITUDE, rho=rho, snr=snr_fixed),
                         support=None, s=S),
        DA_SPEC, K,
        hdr_base + ", support=Random (redrawn per trial)",
        include_base_trex=True)


# ==============================================================================
# Part 3: 2D Gap x Rho sweep
#
#  Runs num_MC trials per (solver, gap, rho) cell for two support types:
#    A) CappedSpread(gap) — deterministic, evenly-spaced support.
#    B) Random            — support redrawn each trial (not gap-dependent).
#
#  The kappa boundary (kappa = ceil(log(0.02) / log(rho))) is the DA+AR1
#  correction window half-width. When gap < kappa, active predictors fall
#  inside each other's correction windows and TPP collapses.
# ==============================================================================

def part_3_gap_rho_sweep():
    snr_fixed = 2.0
    gap_grid = [100, 50, 20, 15, 10, 5, 1]
    rho_grid = [round(0.1 * i, 1) for i in range(10)]

    kappa_boundary = [0 if rho <= 0.0
                      else int(np.ceil(np.log(0.02) / np.log(rho)))
                      for rho in rho_grid]

    solvers = default_solvers("+AR1")
    names = [sv["name"] for sv in solvers]
    n_rho, n_gap = len(rho_grid), len(gap_grid)

    tpp_cs = {nm: np.zeros((n_rho, n_gap)) for nm in names}
    fdp_cs = {nm: np.zeros((n_rho, n_gap)) for nm in names}
    sd_tpp_cs = {nm: np.zeros((n_rho, n_gap)) for nm in names}
    sd_fdp_cs = {nm: np.zeros((n_rho, n_gap)) for nm in names}
    tpp_rand = {nm: np.zeros(n_rho) for nm in names}
    fdp_rand = {nm: np.zeros(n_rho) for nm in names}
    sd_tpp_rand = {nm: np.zeros(n_rho) for nm in names}
    sd_fdp_rand = {nm: np.zeros(n_rho) for nm in names}

    print_section_header("AR(1) Part 3: Gap x Rho sweep")
    print(f"  n={N}  p={P}  s={S}  SNR={snr_fixed}  MC={NUM_MC}"
          f"  tFDR={TFDR}  amplitude={AMPLITUDE}")
    print("  gap_grid: {" + ", ".join(map(str, gap_grid)) + "}")
    print("  rho_grid: {" + ", ".join(f"{r:.1f}" for r in rho_grid) + "}\n")

    # Constant seed offset across cells (mirrors cpp base_offset = base_seed)
    base_offset = BASE_SEED

    for sv in solvers:
        nm = sv["name"]
        print(f"\n  Solver: {nm}\n")

        # ---- Section A: CappedSpread(gap) x rho ----
        print("  [A] CappedSpread(gap) sweep ...\n")
        cell_idx, total_cells = 0, n_rho * n_gap
        for ir, rho in enumerate(rho_grid):
            for ig, gap in enumerate(gap_grid):
                cell_idx += 1
                support = make_support_capped_spread(S, P, gap)
                lbl = (f"[{cell_idx}/{total_cells}] {nm}  rho={rho:.1f}"
                       f"  gap={gap:3d}  kappa={kappa_boundary[ir]}")
                res = run_mc_trials(
                    NUM_MC, lbl, N, P, "dgp_ar1_snr",
                    dict(amplitude=AMPLITUDE, rho=rho, snr=snr_fixed),
                    support, S, TFDR, sv, K, DA_SPEC, base_offset)
                tpp_cs[nm][ir, ig] = res["avg_tpr"]
                fdp_cs[nm][ir, ig] = res["avg_fdr"]
                sd_tpp_cs[nm][ir, ig] = res["sd_tpr"]
                sd_fdp_cs[nm][ir, ig] = res["sd_fdr"]

        # ---- Section B: Random support sweep (rho only) ----
        print("\n  [B] Random support sweep ...\n")
        for ir, rho in enumerate(rho_grid):
            lbl = (f"[{ir + 1}/{n_rho}] {nm}  rho={rho:.1f}"
                   f"  kappa={kappa_boundary[ir]}  Random")
            res = run_mc_trials(
                NUM_MC, lbl, N, P, "dgp_ar1_snr",
                dict(amplitude=AMPLITUDE, rho=rho, snr=snr_fixed),
                None, S, TFDR, sv, K, DA_SPEC, base_offset)
            tpp_rand[nm][ir] = res["avg_tpr"]
            fdp_rand[nm][ir] = res["avg_fdr"]
            sd_tpp_rand[nm][ir] = res["sd_tpr"]
            sd_fdp_rand[nm][ir] = res["sd_fdr"]

    header_extra = (f"n={N}, p={P}, s={S}, SNR={fmt_num(snr_fixed)}"
                    f", tFDR={fmt_num(TFDR)}, amplitude={fmt_num(AMPLITUDE)}")
    save_and_print_2d_gap_rho_results(
        OUT_DIR, "da_ar1_gap_rho", gap_grid, rho_grid, kappa_boundary, NUM_MC,
        names, tpp_cs, fdp_cs, sd_tpp_cs, sd_fdp_cs,
        tpp_rand, fdp_rand, sd_tpp_rand, sd_fdp_rand, header_extra)


# ==============================================================================
# Main
# ==============================================================================

if __name__ == "__main__":
    if RUN_PART_1:
        part_1_snr_sweep()
    if RUN_PART_2:
        part_2_rho_sweep()
    if RUN_PART_3:
        part_3_gap_rho_sweep()
    print("\nAR(1) MC simulation complete.")
